src/plugins/hidsim/qmp_connection.h

```diff
@@ -0,0 +1,151 @@
+* This also allows you to audit the software for security holes.          *
+*                                                                         *
+* Source code also allows you to port DRAKVUF to new platforms, fix bugs, *
+* and add new features.  You are highly encouraged to submit your changes *
+* on https://github.com/tklengyel/drakvuf, or by other methods.           *
+* By sending these changes, it is understood (unless you specify          *
+* otherwise) that you are offering unlimited, non-exclusive right to      *
+* reuse, modify, and relicense the code.  DRAKVUF will always be          *
+* available Open Source, but this is important because the inability to   *
+* relicense code has caused devastating problems for other Free Software  *
+* projects (such as KDE and NASM).                                        *
+* To specify special license conditions of your contributions, just say   *
+* so when you send them.                                                  *
+*                                                                         *
+* This program is distributed in the hope that it will be useful, but     *
+* WITHOUT ANY WARRANTY; without even the implied warranty of              *
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+* license file for more details (it's in a COPYING file included with     *
+* DRAKVUF, and also available from                                        *
+* https://github.com/tklengyel/drakvuf/COPYING)                           *
+*                                                                         *
+* This file was created by Jan Gruber.                                    *
+* It is distributed as part of DRAKVUF under the same license             *
+***************************************************************************/
+
+#ifndef QMP_H
+#define QMP_H
+
+#include <sys/un.h>
+#include <json-c/json.h>
+#include <json-c/json_object.h>
+#include <stdbool.h>
+
+/* QMP command string templates */
+#define QMP_EXEC_CAPABILITIES "{ \"execute\": \"qmp_capabilities\" }"
+#define QMP_SEND_INPUT_OPENING "{ \"execute\": \"input-send-event\" , \"arguments\": { \"events\": ["
+#define QMP_MOUSE_MOVE_EVENT_FMT_STR " { \"type\": \"%s\", \"data\" : {\"axis\": \"x\", \"value\": %d } }, {\"type\": \"%s\", \"data\": { \"axis\": \"y\", \"value\": %d } }"
+#define QMP_MOUSE_BTN_EVENT_FMT_STR " { \"type\": \"btn\" , \"data\" : { \"down\": %s, \"button\": \"%s\" } } %c"
+#define QMP_SCREEN_DUMP_FMT_STR "{ \"execute\":\"screendump\", \"arguments\": { \"filename\": \"%s\" } }"
+#define QMP_KEY_PRESS_QAPI_FMT_STR " { \"type\": \"key\", \"data\" : { \"down\": %s, \"key\": {\"type\": \"qcode\", \"data\": \"%s\" } } } %c "
+#define QMP_KEY_PRESS_CODE_FMT_STR " { \"type\": \"key\", \"data\" : { \"down\": %s, \"key\": {\"type\": \"number\", \"data\": %d } } } %c "
+#define QMP_SEND_INPUT_CLOSING "] } }"
+
+#define QMP_SUCCESS "{\"return\": {}}\r\n"
```

Comment:
IMO it wouldn't hurt if the commands were generated with json-c.

Reply:
I though about that, but what would the benefit beside from notable overhead? 
A lot of events get generated in a microsecond intervals. It just adds complexity...especially computional-wise.

Reply:
Sound like a premature optimization to me.
We still have to send the data over the socket, QEMU has to parse JSON etc. Quite a lot of overhead comparing to what we have to do here.

Safety and maintainability are IMO more important than raw performance (at least at this stage).

Reply:
Sure, maintainability comes first. But I think the maintainability of those simple format-strings is at least as good as the construction of json-objects IMHO. If you insist, I'll build up those strings with json-c...just say

Reply:
I agree with @jgru, fairly simple for now, no issue with keeping it as-is

Reply:
We have a lot of problems with JSON being glued "by hand" already.

I think I've submitted at least 10 PRs correcting typos in such JSONs generated "by hand" and this was pretty frustrating as the JSON parse errors are thrown in runtime.

Right now some user-provided parameters are glued with JSON template using format string, this was a source of errors in the past already.
https://github.com/tklengyel/drakvuf/pull/1245/files#r664170005

Reply:
I changed each and every QMP-command construction inside hid_injection.cpp to be based on json_objects.